packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart

```diff
@@ -293,19 +380,32 @@ abstract class RenderSliverMultiBoxAdaptor extends RenderSliver
 
   /// Called after layout with the number of children that can be garbage
   /// collected at the head and tail of the child list.
+  ///
+  /// Children whose [SliverMultiBoxAdaptorParentData.keepAlive] property is
+  /// set to true will be removed to a cache instead of being dropped.
+  ///
+  /// This method also collects any children that were previously kept alive but
+  /// are now no longer necessary. As such, it should be called every time
+  /// [performLayout] is run, even if the arguments are both zero.
   @protected
   void collectGarbage(int leadingGarbage, int trailingGarbage) {
     assert(_debugAssertChildListLocked());
     assert(childCount >= leadingGarbage + trailingGarbage);
     invokeLayoutCallback<SliverConstraints>((SliverConstraints constraints) {
       while (leadingGarbage > 0) {
-        _childManager.removeChild(firstChild);
+        _destroyOrCacheChild(firstChild);
         leadingGarbage -= 1;
       }
       while (trailingGarbage > 0) {
-        _childManager.removeChild(lastChild);
+        _destroyOrCacheChild(lastChild);
         trailingGarbage -= 1;
       }
+      _keepAliveBucket.values.where((RenderBox child) {
+        final SliverMultiBoxAdaptorParentData childParentData = child.parentData;
+        return !childParentData.keepAlive;
+      }).toList().forEach((RenderBox child) {
```

Comment:
Yes, removeChild very much modifies _keepAliveBucket, that's the purpose of the call in fact. :-)